File: 09_tries/tries.ts
/**
 * TRIES  ·  TypeScript
 * Trie, WildcardTrie, autocomplete, replace words, word search II.
 */

class TrieNode {
  children: Map<string, TrieNode> = new Map();
  isEnd = false;
  count = 0;
}

class Trie {
  root = new TrieNode();

  insert(word: string): void {
    let node = this.root;
    for (const c of word) {
      if (!node.children.has(c)) node.children.set(c, new TrieNode());
      node = node.children.get(c)!; node.count++;
    }
    node.isEnd = true;
  }

  search(word: string): boolean {
    let node = this.root;
    for (const c of word) { if (!node.children.has(c)) return false; node = node.children.get(c)!; }
    return node.isEnd;
  }

  startsWith(prefix: string): boolean {
    let node = this.root;
    for (const c of prefix) { if (!node.children.has(c)) return false; node = node.children.get(c)!; }
    return true;
  }

  autocomplete(prefix: string): string[] {
    let node = this.root;
    for (const c of prefix) { if (!node.children.has(c)) return []; node = node.children.get(c)!; }
    const results: string[] = [];
    const dfs = (n: TrieNode, cur: string) => {
      if (n.isEnd) results.push(cur);
      for (const [c, child] of [...n.children.entries()].sort()) dfs(child, cur+c);
    };
    dfs(node, prefix); return results;
  }

  delete(word: string): boolean {
    const del = (node: TrieNode, i: number): boolean => {
      if (i === word.length) { if (!node.isEnd) return false; node.isEnd = false; return node.children.size === 0; }
      const c = word[i]; if (!node.children.has(c)) return false;
      const shouldDel = del(node.children.get(c)!, i+1);
      if (shouldDel) node.children.delete(c);
      return shouldDel && !node.isEnd && node.children.size === 0;
    };
    return del(this.root, 0);
  }
}

class WildcardTrie {
  root = new TrieNode();
  insert(word: string): void {
    let node = this.root;
    for (const c of word) { if (!node.children.has(c)) node.children.set(c, new TrieNode()); node = node.children.get(c)!; }
    node.isEnd = true;
  }
  search(word: string): boolean {
    const dfs = (node: TrieNode, i: number): boolean => {
      if (i === word.length) return node.isEnd;
      const c = word[i];
      if (c === '.') return [...node.children.values()].some(child => dfs(child, i+1));
      return node.children.has(c) ? dfs(node.children.get(c)!, i+1) : false;
    };
    return dfs(this.root, 0);
  }
}

// ── Problems ──────────────────────────────────
function replaceWords(dictionary: string[], sentence: string): string {
  const t = new Trie(); dictionary.forEach(r => t.insert(r));
  return sentence.split(' ').map(word => {
    let node = t.root, replacement = '';
    for (const c of word) {
      if (!node.children.has(c)) break;
      node = node.children.get(c)!; replacement += c;
      if (node.isEnd) break;
    }
    return node.isEnd ? replacement : word;
  }).join(' ');
}

function wordSearchII(board: string[][], words: string[]): string[] {
  const t = new Trie(); words.forEach(w => t.insert(w));
  const rows = board.length, cols = board[0].length, found = new Set<string>();
  const dfs = (node: TrieNode, r: number, c: number, path: string) => {
    if (r<0||r>=rows||c<0||c>=cols||!node.children.has(board[r][c])) return;
    const ch = board[r][c]; node = node.children.get(ch)!; path += ch;
    if (node.isEnd) found.add(path);
    board[r][c] = '#';
    for (const [dr,dc] of [[0,1],[0,-1],[1,0],[-1,0]]) dfs(node, r+dr, c+dc, path);
    board[r][c] = ch;
  };
  for (let r=0;r<rows;r++) for (let c=0;c<cols;c++) dfs(t.root, r, c, '');
  return [...found];
}

function longestWordInDictionary(words: string[]): string {
  const t = new Trie(); words.forEach(w => t.insert(w));
  let best = '';
  const dfs = (node: TrieNode, cur: string) => {
    if (cur.length > best.length || (cur.length === best.length && cur < best)) best = cur;
    for (const [c, child] of [...node.children.entries()].sort()) if (child.isEnd) dfs(child, cur+c);
  };
  dfs(t.root, ''); return best;
}

// ── Tests ─────────────────────────────────────
function assert(c: boolean, m: string) { if (!c) throw new Error(`FAIL: ${m}`); }
const eq = (a: unknown, b: unknown) => JSON.stringify(a) === JSON.stringify(b);

function runTests() {
  console.log("Running trie tests...\n");

  const t = new Trie();
  ["apple","app","application","apply"].forEach(w => t.insert(w));
  assert(t.search("apple")&&t.search("app"),"search found");
  assert(!t.search("ap")&&!t.search("apples"),"search not found");
  assert(t.startsWith("app")&&!t.startsWith("xyz"),"startsWith");
  console.log("  ✅ Trie: insert / search / startsWith");

  assert(eq(t.autocomplete("app"),["app","apple","application","apply"]),"autocomplete");
  console.log("  ✅ Trie: autocomplete");

  t.delete("app");
  assert(!t.search("app")&&t.search("apple"),"delete");
  console.log("  ✅ Trie: delete");

  const wt = new WildcardTrie();
  ["bad","dad","mad"].forEach(w => wt.insert(w));
  assert(wt.search("bad")&&wt.search(".ad")&&wt.search("b.."),"wildcard match");
  assert(!wt.search("pad")&&!wt.search("ba"),"wildcard no match");
  console.log("  ✅ WildcardTrie: exact / wildcard / mismatch");

  assert(replaceWords(["cat","bat","rat"],"the cattle was rattled by the battery")==="the cat was rat by the bat","replaceWords");
  console.log("  ✅ replaceWords");

  const board=[["o","a","a","n"],["e","t","a","e"],["i","h","k","r"],["i","f","l","v"]];
  const found=new Set(wordSearchII(board,["oath","pea","eat","rain"]));
  assert(found.has("oath")&&found.has("eat"),"wordSearch");
  console.log("  ✅ wordSearchII");

  assert(longestWordInDictionary(["w","wo","wor","worl","world"])==="world","longest");
  console.log("  ✅ longestWordInDictionary");

  console.log("\n🎉 All trie tests passed!");
}
runTests();
export { TrieNode, Trie, WildcardTrie, replaceWords, wordSearchII, longestWordInDictionary };

File: 09_tries/tries.py
"""
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
TRIES  ·  Python
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
Trie (prefix tree): each node = one character, path = prefix.
Insert/search/startsWith: O(L) where L = word length.
Space: O(ALPHABET_SIZE * L * N). Best for prefix queries, autocomplete.
"""


class TrieNode:
    def __init__(self):
        self.children: dict[str, 'TrieNode'] = {}
        self.is_end = False
        self.count = 0      # words passing through this node


class Trie:
    def __init__(self): self.root = TrieNode()

    def insert(self, word: str) -> None:
        node = self.root
        for c in word:
            if c not in node.children: node.children[c] = TrieNode()
            node = node.children[c]; node.count += 1
        node.is_end = True

    def search(self, word: str) -> bool:
        node = self.root
        for c in word:
            if c not in node.children: return False
            node = node.children[c]
        return node.is_end

    def starts_with(self, prefix: str) -> bool:
        node = self.root
        for c in prefix:
            if c not in node.children: return False
            node = node.children[c]
        return True

    def delete(self, word: str) -> bool:
        def _del(node: TrieNode, word: str, i: int) -> bool:
            if i == len(word):
                if not node.is_end: return False
                node.is_end = False; return len(node.children)==0
            c = word[i]
            if c not in node.children: return False
            should_delete = _del(node.children[c], word, i+1)
            if should_delete: del node.children[c]
            return should_delete and not node.is_end and len(node.children)==0
        return _del(self.root, word, 0)

    def autocomplete(self, prefix: str) -> list[str]:
        node = self.root
        for c in prefix:
            if c not in node.children: return []
            node = node.children[c]
        results = []
        def dfs(n, cur):
            if n.is_end: results.append(cur)
            for c, child in sorted(n.children.items()): dfs(child, cur+c)
        dfs(node, prefix); return results

    def count_words_with_prefix(self, prefix: str) -> int:
        node = self.root
        for c in prefix:
            if c not in node.children: return 0
            node = node.children[c]
        return node.count


# ══════════════════════════════════════
# ── WildcardTrie class ──────────────────────
# ══════════════════════════════════════

class WildcardTrie:
    """Trie supporting '.' as wildcard (matches any single character)."""
    def __init__(self): self.root = TrieNode()

    def insert(self, word: str):
        node = self.root
        for c in word:
            if c not in node.children: node.children[c] = TrieNode()
            node = node.children[c]
        node.is_end = True

    def search(self, word: str) -> bool:
        def dfs(node: TrieNode, i: int) -> bool:
            if i == len(word): return node.is_end
            c = word[i]
            if c == '.':
                return any(dfs(child, i+1) for child in node.children.values())
            if c not in node.children: return False
            return dfs(node.children[c], i+1)
        return dfs(self.root, 0)


# ══════════════════════════════════════
# ── Problems ──────────────────────
# ══════════════════════════════════════

def word_search_ii(board: list[list[str]], words: list[str]) -> list[str]:
    """🔴 Word Search II (LC #212) — Trie + DFS backtracking"""
    trie = Trie()
    for w in words: trie.insert(w)
    rows, cols = len(board), len(board[0])
    found = set()

    def dfs(node: TrieNode, r: int, c: int, path: str):
        if r<0 or r>=rows or c<0 or c>=cols or board[r][c] not in node.children: return
        ch = board[r][c]; node = node.children[ch]; path += ch
        if node.is_end: found.add(path)
        board[r][c] = '#'
        for dr,dc in [(0,1),(0,-1),(1,0),(-1,0)]: dfs(node, r+dr, c+dc, path)
        board[r][c] = ch

    for r in range(rows):
        for c in range(cols): dfs(trie.root, r, c, "")
    return list(found)

def replace_words(dictionary: list[str], sentence: str) -> str:
    """🟡 Replace Words (LC #648) — replace with shortest root"""
    trie = Trie()
    for root in dictionary: trie.insert(root)
    result = []
    for word in sentence.split():
        node = trie.root; replacement = ""
        for c in word:
            if c not in node.children: break
            node = node.children[c]; replacement += c
            if node.is_end: break
        result.append(replacement if node.is_end else word)
    return " ".join(result)

def longest_word_dictionary(words: list[str]) -> str:
    """🟡 Longest Word in Dictionary (LC #720)"""
    trie = Trie()
    for w in words: trie.insert(w)
    best = ""
    def dfs(node: TrieNode, cur: str):
        nonlocal best
        if len(cur)>len(best) or (len(cur)==len(best) and cur<best): best=cur
        for c in sorted(node.children):
            child = node.children[c]
            if child.is_end: dfs(child, cur+c)
    dfs(trie.root, ""); return best

def palindrome_pairs(words: list[str]) -> list[list[int]]:
    """🔴 Palindrome Pairs (LC #336) — hash map approach"""
    lookup = {w:i for i,w in enumerate(words)}
    result = []
    def is_pal(s): return s==s[::-1]
    for i, w in enumerate(words):
        for j in range(len(w)+1):
            pre, suf = w[:j], w[j:]
            if is_pal(pre):
                rev = suf[::-1]
                if rev != w and rev in lookup: result.append([lookup[rev],i])
            if j < len(w) and is_pal(suf):
                rev = pre[::-1]
                if rev != w and rev in lookup: result.append([i,lookup[rev]])
    return result

def find_max_xor(nums: list[int]) -> int:
    """🟡 Maximum XOR of Two Numbers (LC #421) — bit trie"""
    max_xor = 0; prefix = 0
    for mask in range(31, -1, -1):
        prefix |= (1 << mask)
        prefixes = set(n & prefix for n in nums)
        candidate = max_xor | (1 << mask)
        if any((candidate ^ p) in prefixes for p in prefixes):
            max_xor = candidate
    return max_xor

# ══════════════════════════════════════
# ── Tests ──────────────────────
# ══════════════════════════════════════

def run_tests():
    print("Running trie tests...\n")

    t = Trie()
    for w in ["apple","app","application","apply"]: t.insert(w)
    assert t.search("apple") and t.search("app")
    assert not t.search("ap") and not t.search("apples")
    assert t.starts_with("app") and t.starts_with("appl")
    assert not t.starts_with("xyz")
    print("  ✅ Trie: insert / search / starts_with")

    assert t.autocomplete("app") == ["app","apple","application","apply"]
    assert t.autocomplete("xyz") == []
    print("  ✅ Trie: autocomplete")

    t.delete("app")
    assert not t.search("app") and t.search("apple")
    print("  ✅ Trie: delete (removes word, keeps prefixes)")

    wt = WildcardTrie()
    for w in ["bad","dad","mad"]: wt.insert(w)
    assert wt.search("bad") and wt.search(".ad") and wt.search("b..")
    assert not wt.search("pad") and not wt.search("ba")
    print("  ✅ WildcardTrie: exact / wildcard / mismatch")

    assert replace_words(["cat","bat","rat"],"the cattle was rattled by the battery") \
        == "the cat was rat by the bat"
    print("  ✅ replace_words")

    assert longest_word_dictionary(["w","wo","wor","worl","world"]) == "world"
    assert longest_word_dictionary(["a","banana","app","appl","ap","apply","apple"]) == "apple"
    print("  ✅ longest_word_dictionary")

    board = [["o","a","a","n"],["e","t","a","e"],["i","h","k","r"],["i","f","l","v"]]
    found = set(word_search_ii(board, ["oath","pea","eat","rain"]))
    assert "oath" in found and "eat" in found
    print("  ✅ word_search_ii")

    pairs = palindrome_pairs(["abcd","dcba","lls","s","sssll"])
    assert [0,1] in pairs and [1,0] in pairs
    print("  ✅ palindrome_pairs")

    assert find_max_xor([3,10,5,25,2,8]) == 28
    print("  ✅ find_max_xor")

    print("\n🎉 All trie tests passed!")

if __name__ == "__main__":
    run_tests()
